build-scripts/graal.js

```diff
@@ -0,0 +1,211 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+/**
+ * @fileoverview
+ *
+ * Build the graal native compiler image for the current OS.
+ * Intended to be run with a working directory of the intended package.
+ */
+
+const {spawn} = require('child_process');
+const fs = require('fs');
+const path = require('path');
+const ncp = require('ncp');
+const {
+  GRAAL_OS,
+  GRAAL_FOLDER,
+  GRAAL_VERSION,
+  GRAAL_URL
+} = require('./graal-env');
+const TEMP_PATH = path.resolve(__dirname, '../temp');
+
+process.on('unhandledRejection', error => {
+  console.error(error);
+  process.exit(1);
+});
+
+// The latest compiler is now compatible with a released version of graal
+const GRAAL_BUILD_FROM_SRC = !Boolean('COMPILER_NIGHTLY' in process.env);
+
+/**
+ * Simple wrapper for the NodeJS spawn command to use promises
+ *
+ * @param {string} command
+ * @param {Object=} spawnOpts
+ * @return {!Promise}
+ */
+function runCommand(command, spawnOpts = {}) {
+  process.stdout.write(`${command}\n`);
+  return new Promise((resolve, reject) => {
+    const commandParts = command.split(/\s+/);
+    const opts = Object.assign({}, {
+      stdio: 'inherit'
+    }, spawnOpts);
+    const externalProcess = spawn(commandParts[0], commandParts.slice(1), opts);
+    externalProcess.on('error', err => {
+      process.stderr.write(`${err.toString()}\n`);
+      reject(err);
+    });
+    externalProcess.on('close', exitCode => {
+      if (exitCode != 0) {
+        process.stderr.write(`non zero exit code ${exitCode}\n`);
+        process.exit(1);
+        reject();
+      }
+      resolve();
+    });
+  });
+}
+
+// Build graal from source
+if (!fs.existsSync(TEMP_PATH)) {
+  fs.mkdirSync(TEMP_PATH);
+}
+
+const NATIVE_IMAGE_BUILD_ARGS = [
+  '-H:+JNI',
+  '--no-server',
+  '-H:+ReportUnsupportedElementsAtRuntime',
+  '-H:IncludeResourceBundles=com.google.javascript.rhino.Messages',
+  '-H:IncludeResourceBundles=org.kohsuke.args4j.Messages',
+  '-H:IncludeResourceBundles=org.kohsuke.args4j.spi.Messages',
+  '-H:IncludeResourceBundles=com.google.javascript.jscomp.parsing.ParserConfig',
+  `-H:ReflectionConfigurationFiles=${path.resolve(__dirname, 'reflection-config.json')}`,
+  '-H:IncludeResources="(externs.zip)|(.*(js|txt))"',
+  '-jar',
+  path.resolve(process.cwd(), 'compiler.jar')
+];
+let buildSteps = Promise.resolve();
+if (GRAAL_BUILD_FROM_SRC) {
+  const GRAAL_SRC_BASE = path.resolve(TEMP_PATH, 'graal');
+  const MX_SRC_BASE = path.resolve(TEMP_PATH, 'mx');
+  const JVMCI_VERSION = 'jvmci-0.44';
+
+  // Clone the graal and mx repositories
+  buildSteps = buildSteps
```

Comment:
Could you indicate the expected state of the temporary directory after this step?
e.g.
// After this step TEMP_PATH should contain:
//   graal/
//   mx/

Same for later steps that add things to the temporary directory